### PR TITLE
Replace PooledStopwatch with SharedStopwatch

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
What was the benefit here?  Just to have less stuff to track through a pool?

**Comment by sharwell** [comment]
The pools are not particularly efficient on many-core machines. In this case, it's trivial to implement the functionality without any allocations at all.

File: src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+namespace Roslyn.Utilities
+{
+    internal struct SharedStopwatch
+    {
+        private static readonly Stopwatch s_stopwatch = Stopwatch.StartNew();
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs]: There could be issues if the program is run for more than [10 million days](https://docs.microsoft.com/en-us/dotnet/api/system.timespan.maxvalue?view=netframework-4.8#remarks), but this seems like an unusual scenario.

**Comment by CyrusNajmabadi** [comment]
> The pools are not particularly efficient on many-core machines. In this case, it's trivial to implement the functionality without any allocations at all.

SGTM.

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler  for reviews

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for a second review

File: src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public void Stop()
+        {
+            if (IsRunning)
+            {
+                _accumulated += s_stopwatch.Elapsed - _started;
+                _started = TimeSpan.Zero;
+            }
+        }
```
Comment (cston) [inline src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs]: Are these methods needed? It looks like the only method used is `Start()` above which could be replaced by `_started = ...;`.

File: src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj [hunk 3]
```diff
@@ -27,6 +27,7 @@
   <ItemGroup Label="Linked Files">
     <Compile Remove="Storage\Sqlite\**\*.cs" Condition="'$(DotNetBuildFromSource)' == 'true'" />
     <Compile Include="..\..\..\CodeStyle\Core\CodeFixes\FixAllContextHelper.cs" Link="CodeFixes\FixAllOccurrences\FixAllContextHelper.cs" />
+    <Compile Include="..\..\..\Compilers\Core\Portable\InternalUtilities\SharedStopwatch.cs" Link="InternalUtilities\SharedStopwatch.cs" />
```
Comment (AlekseyTs) [inline src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj]: ><Compile Include="..\..\..\Compilers\Core\Portable\InternalUtilities\SharedStopwatch.cs" Link="InternalUtilities\SharedStopwatch.cs" /> [](start = 4, length = 135)

It looks like ```AnalyzerExecutor``` is the only consumer of the new type and it is not linked explicitly by this project. Are there other consumers I missed?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj]: There are no current consumers, but I am planning to replace some other uses of `Stopwatch` with this type for improved performance.

File: src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs]: I was stuck deciding between drop-in full functionality and limited functionality for scenarios most impacted by allocation performance. There didn't seem to be much downside to the former considering the simplicity of the API.

Comment (cston) [reply src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs]: We should keep it simple. Could the type could be reduced to two `static` methods?
```C#
static long StartNew() => s_stopwatch.ElapsedTicks;

static long Stop(long start) => s_stopwatch.ElapsedTicks - start;
```

---
In reply to: [371412091](https://github.com/dotnet/roslyn/pull/41205#discussion_r371412091) [](ancestors = 371412091)

File: src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj (hunk 3 above)
Comment (AlekseyTs) [reply src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj]: > There are no current consumers, but I am planning to replace some other uses of Stopwatch with this type for improved performance.

I think the project file should be changed in the same PR that makes use of the new type in the project.

---
In reply to: [371411528](https://github.com/dotnet/roslyn/pull/41205#discussion_r371411528) [](ancestors = 371411528)

File: src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs (hunk 2 above)
Comment (tannergooding) [reply src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs]: Stopwatch itself exposes static APIs that return the raw tick count and frequency
in which case you don't even need a stopwatch instance, you just need to normalize to 100ns ticks

Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/SharedStopwatch.cs]: Reduced the API to just `StartNew()` and `Elapsed`

File: src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj]: Removed the unnecessary project change

**Comment by sharwell** [comment]
@AlekseyTs @cston I addressed both of your feedback now

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs [hunk 4]
```diff
@@ -1448,32 +1448,31 @@ private void ExecuteAndCatchIfThrows_NoLock<TArg>(DiagnosticAnalyzer analyzer, A
...
+                    // call to StartNew blocked on a GC operation, this call to StartNew will most likely occur when the
                     // GC is no longer active. In practice, a substantial improvement to the consistency of analyzer
                     // timing data was observed.
                     //
-                    // Note that the call to Stopwatch.Stop() is not affected, because the GC wait will occur after the
-                    // timer has already recorded its stop time.
-                    timer.Restart();
+                    // Note that the call to SharedStopwatch.Elapsed is not affected, because the GC wait will occur
+                    // after the timer has already recorded its stop time.
+                    timer = SharedStopwatch.StartNew();
```
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs]: >timer = SharedStopwatch.StartNew(); [](start = 20, length = 35)

Is the second call to `StartNew()` necessary?

Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs]: Yes, the comment explains it

**Pull request merged**